### PR TITLE
fix(aiLabel): link on top of  popover fails

### LINKED ISSUE #21235: [Bug]: AILabel popover links fail to open when component is rendered inside Shadow DOM
### Package

@carbon/react

### Browser

All

### Package version

1.49.0

### React version

_No response_

### Description

Links inside Carbon Design System's `AILabel` popover content do not work when the component is rendered within a Shadow DOM web component. The popover closes instead of redirecting to that link.

### Environment
- **Carbon React Version**: 1.49.0
- **Framework**: Preact 10.19.3

## 🎯 Affected Components

This issue affects any interactive content inside AILabel popovers when used in Shadow DOM:
- Links (`<a>` tags)
- Buttons with click handlers
- Any element with `onClick` handlers (such as ai label action buttons as well)

### Reproduction/example

https://stackblitz.com/edit/vitejs-vite-dbhyutku?file=index.html

### Steps to reproduce

1. Create a web component using `preact-custom-element` with `shadow: true`
2. Render an `AILabel` component inside the shadow DOM
3. Add links (`<a>` tags) inside `AILabelContent`
4. Click on any link in the popover
5. **Expected**: Link opens in new tab
6. **Actual**: Popover closes immediately, link does not open

### Suggested Severity

None

### Application/PAL

_No response_

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)
- [x] I checked the [current issues](https://github.com/carbon-design-system/carbon/issues) for duplicate problems

### PR DESCRIPTION
**Pull request by sangeethababu9223**
Closes #21235

Fixes AILabel popover links failing to open when the component is rendered inside Shadow DOM by improving click detection logic in Toggletip.

### Changelog

**Changed**

- Enhanced Toggletip's outside click detection to properly handle Shadow DOM contexts using composedPath() API
- Updated click event handling in Toggletip to check the full event path through shadow boundaries, not just the retargeted event.target
- Updated click event handling in Toggletip to check the full event path through shadow boundaries, not just the retargeted event.target

#### Testing / Reviewing

- Navigate to the Shadow DOM Test story in Storybook
- Test inside click scenario: Click on links and buttons inside the Toggletip content - they should work correctly and the Toggletip should remain open
- Test outside click scenario: Click outside the Toggletip - it should close properly
- Verify the Shadow DOM test suite passes

**Note: The Shadow DOM Test story (AILabelShadowDemo) was added specifically for manual testing purposes and will be removed after review. While using React AILabel on top of Shadow DOM may seem like an edge case (and the Web Components AILabel would typically be preferred), there are valid scenarios where supporting this use case is beneficial, hence this solution.**

Ignore the changes/additions in:
``
packages/react/src/components/AILabel/AILabel.stories.js
packages/react/src/components/AILabel/stories/AILabelShadowDemo.js
``
as they're only for testing purposes and will be reverted.

## PR Checklist

As the author of this PR, before marking ready for review, confirm you:

- [x] Reviewed every line of the diff
- [ ]~ Updated documentation and storybook examples~
- [x] Wrote passing tests that cover this change
- [ ]~Addressed any impact on accessibility (a11y)~
- [x] Tested for cross-browser consistency
- [x] Validated that this code is ready for review and status checks should pass

More details can be found in the [pull request guide](https://github.com/carbon-design-system/carbon/blob/main/docs/guides/reviewing-pull-requests.md)

### CODE REVIEW

File: packages/react/src/components/AILabel/stories/AILabelShadowDemo.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
+/**
+ * Copyright IBM Corp. 2024, 2025
```
Comment (adamalston) [inline packages/react/src/components/AILabel/stories/AILabelShadowDemo.js]: Could you update the years?

File: packages/react/src/components/Toggletip/__tests__/Toggletip-test.js [hunk 2]
```diff
@@ -514,6 +514,115 @@ describe('Toggletip', () => {
...
+            </ToggletipContent>
+          </Toggletip>
+        );
+
+        const innerContent = screen.getByTestId('inner-content');
+        const toggletip = screen.getByTestId('toggletip');
+
+        // Simulate Shadow DOM: event.target is retargeted to shadow host,
+        // but composedPath() contains the full lineage including the actual click target
+        const mockEvent = new MouseEvent('mousedown', { bubbles: true });
```
Comment (adamalston) [inline packages/react/src/components/Toggletip/__tests__/Toggletip-test.js]: This code uses a synthetic `mousedown` event. In browsers that support `PointerEvent`, the handler listens for `pointerdown`, which suggests that the actual event through a `ShadowRoot` remains uncovered. Should coverage be added for that, or do you think I'm missing something?

File: packages/react/src/components/AILabel/stories/AILabelShadowDemo.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    // Inject Carbon styles into shadow root
+    const link = document.createElement('link');
+    link.rel = 'stylesheet';
+    link.href = 'https://unpkg.com/@carbon/styles@1.103.0/css/styles.min.css';
+    this.shadowRoot.appendChild(link);
+
+    // Render React component
+    const root = this.shadowRoot.querySelector('#react-root');
+    const reactRoot = createRoot(root);
```
Comment (adamalston) [inline packages/react/src/components/AILabel/stories/AILabelShadowDemo.js]: Should `reactRoot` be unmounted from `disconnectedCallback`?

File: packages/react/src/components/Toggletip/index.tsx [hunk 4]
```diff
@@ -181,7 +181,22 @@ export function Toggletip<E extends ElementType = 'span'>({
...
+            return ref.current?.contains(element);
+          }
+          return false;
+        });
+      }
+
+      if (open && !isInside) {
         setOpen(false);
       }
     };
```
Comment (adamalston) [inline packages/react/src/components/Toggletip/index.tsx:196]: Suggestion to simplify this function and make it type safe to avoid the need for type assertions.

```tsx
const handleOutsideClick = (event: MouseEvent | PointerEvent) => {
  if (!ref.current) return;

  const { current } = ref;

  const path =
    typeof event.composedPath === 'function' ? event.composedPath() : [];
  const isInsideCurrent = (target: EventTarget | null) =>
    target instanceof Node && current.contains(target);
  // `event.target` is retargeted at Shadow DOM boundaries, so fall back to
  // `composedPath()`.
  const isInside =
    isInsideCurrent(event.target) || path.some(isInsideCurrent);

  if (open && !isInside) {
    setOpen(false);
  }
};
```

File: packages/react/src/components/AILabel/stories/AILabelShadowDemo.js (hunk 1 above)
Comment (sangeethababu9223) [reply packages/react/src/components/AILabel/stories/AILabelShadowDemo.js]: Hey @adamalston ,
You could ignore the changes in this file as this was added purely for testing purposes and will be removed once code reviews are completed.

File: packages/react/src/components/AILabel/stories/AILabelShadowDemo.js (hunk 3 above)
Comment (sangeethababu9223) [reply packages/react/src/components/AILabel/stories/AILabelShadowDemo.js]: Hey @adamalston,
I'll be removing this file once the PR reviews are complete. This was added purely for testing purpose. You could ignore this.

File: packages/react/src/components/Toggletip/__tests__/Toggletip-test.js (hunk 2 above)
Comment (sangeethababu9223) [reply packages/react/src/components/Toggletip/__tests__/Toggletip-test.js]: Good catch. In the test environment (jsdom), PointerEvent isn’t available, so the code falls back to mousedown. Since the handler doesn’t branch on event type and relies on target and composedPath(), the current tests still validate the core behavior, including the Shadow DOM path.

File: packages/react/src/components/Toggletip/index.tsx (hunk 4 above)
Comment (sangeethababu9223) [reply packages/react/src/components/Toggletip/index.tsx:196]: Hey @adamalston ,
Thanks for the suggestion! I've updated the function to use a type predicate for cleaner, assertion-free type safety. I also kept the composedPath() check conditional so it only runs if the initial event.target check fails, which helps keep the logic optimized while still supporting Shadow DOM.

**Review by maradwan26 — APPROVED** [review]
LGTM!

File: packages/react/src/components/Toggletip/__tests__/Toggletip-test.js [hunk 5]
```diff
@@ -514,6 +514,101 @@ describe('Toggletip', () => {
...
+              <div data-testid="inner-content">Content</div>
+            </ToggletipContent>
+          </Toggletip>
+        );
+
+        const innerContent = screen.getByTestId('inner-content');
+        const toggletip = screen.getByTestId('toggletip');
+
+        const mockEvent = new MouseEvent('mousedown', { bubbles: true });
+        delete mockEvent.composedPath;
```
Comment (adamalston) [inline packages/react/src/components/Toggletip/__tests__/Toggletip-test.js]: Could you override `composedPath` to `undefined` here instead of deleting it? Since the method is inherited from `Event.prototype`, deleting it from the event instance still leaves it available, so the fallback branch that relies on `event.target` doesn't appear to actually be tested.

File: packages/react/src/components/Toggletip/__tests__/Toggletip-test.js [hunk 6]
```diff
@@ -514,6 +514,101 @@ describe('Toggletip', () => {
...
+        const innerContent = screen.getByTestId('inner-content');
+        const toggletip = screen.getByTestId('toggletip');
+
+        const mockEvent = new MouseEvent('mousedown', { bubbles: true });
+        Object.defineProperty(mockEvent, 'composedPath', {
+          value: () => [innerContent, toggletip, document.body],
+        });
+        Object.defineProperty(mockEvent, 'target', {
+          value: toggletip,
+        });
```
Comment (adamalston) [inline packages/react/src/components/Toggletip/__tests__/Toggletip-test.js:554]: Should `event.target` be set to a Shadow DOM host here instead of `toggletip`? Because `target` is still a node inside the component, the pre-fix `contains(node)` check would also pass, so this test does not seem to exercise the Shadow DOM retargeting case that the fix is meant to cover.

File: packages/react/src/components/Toggletip/__tests__/Toggletip-test.js (hunk 5 above)
Comment (sangeethababu9223) [reply packages/react/src/components/Toggletip/__tests__/Toggletip-test.js]: Thank you for pointing this out. You're right. I've updated the test case with the suggested change.

File: packages/react/src/components/Toggletip/__tests__/Toggletip-test.js (hunk 6 above)
Comment (sangeethababu9223) [reply packages/react/src/components/Toggletip/__tests__/Toggletip-test.js:554]: I've updated the target with `shadowHost` instead of `toggletip`. This does make more sense from a testing perspective. you're absolutely right.

File: packages/react/src/components/Toggletip/index.tsx [hunk 7]
```diff
@@ -179,9 +179,18 @@ export function Toggletip<E extends ElementType = 'span'>({
...
-      if (open && node && !ref.current?.contains(node)) {
+      const isInsideCurrent = (target: EventTarget | null): target is Node =>
+        target instanceof Node && current.contains(target);
+
+      const isInside =
+        isInsideCurrent(event.target) ||
+        (typeof event.composedPath === 'function' &&
+          event.composedPath().some(isInsideCurrent));
+
+      if (open && !isInside) {
```
Comment (adamalston) [inline packages/react/src/components/Toggletip/index.tsx]: Should this `open` check be moved so that the listener is only attached while the `ToggleTip` is open?

```diff
diff --git a/packages/react/src/components/Toggletip/index.tsx b/packages/react/src/components/Toggletip/index.tsx
index 14759ab7f9..20d33e0176 100644
--- a/packages/react/src/components/Toggletip/index.tsx
+++ b/packages/react/src/components/Toggletip/index.tsx
@@ -172,7 +172,7 @@ export function Toggletip<E extends ElementType = 'span'>({
   });
 
   useEffect(() => {
-    if (!ref.current) return;
+    if (!open || !ref.current) return;
 
     const targetDocument = ref.current.ownerDocument || document;
     const eventType: 'pointerdown' | 'mousedown' =
@@ -190,7 +190,7 @@ export function Toggletip<E extends ElementType = 'span'>({
         (typeof event.composedPath === 'function' &&
           event.composedPath().some(isInsideCurrent));
 
-      if (open && !isInside) {
+      if (!isInside) {
         setOpen(false);
       }
     };
```

File: packages/react/src/components/Toggletip/__tests__/Toggletip-test.js [hunk 8]
```diff
@@ -514,6 +514,133 @@ describe('Toggletip', () => {
         expect(onKeyDown).toHaveBeenCalledTimes(1);
       });
     });
+
+    describe('Shadow DOM Support', () => {
+      // Tests verify handling of event.target retargeting in Shadow DOM using composedPath()
```
Comment (adamalston) [inline packages/react/src/components/Toggletip/__tests__/Toggletip-test.js]: Are the details in this comment redundant with the tests in this `describe` block? If so, should the comment be removed?

**Review by adamalston — COMMENTED** [review]
I'm ready to approve pending removal of `packages/react/src/components/AILabel/stories/AILabelShadowDemo.js`.

File: packages/react/src/components/Toggletip/index.tsx (hunk 7 above)
Comment (sangeethababu9223) [reply packages/react/src/components/Toggletip/index.tsx]: Makes sense, I've updated the conditions. 
Thanks @adamalston.

**Review by devadula-nandan — APPROVED** [review]
